internal/pkg/composable/providers/kubernetes/hints.go

```diff
@@ -250,83 +253,66 @@ func GenerateHintsMapping(hints mapstr.M, kubeMeta mapstr.M, logger *logp.Logger
-		annotations, _ := ann.(mapstr.M)
-
-		if containerEntries, err := annotations.GetValue(prefix + ".hints"); err == nil {
-			entries, ok := containerEntries.(mapstr.M)
-			if ok && len(entries) > 0 {
-				for key := range entries {
-					parts := strings.Split(key, "/")
-
-					if len(parts) > 1 {
-						if con, ok := k8sMapping["container"]; ok {
-							containers, ok := con.(mapstr.M)
-							if ok {
-								if cname, err := containers.GetValue("name"); err == nil {
-									if parts[0] == cname {
-										// If there are hints like co.elastic.hints.<container_name>/ then add the values after the / to the corresponding container
-										// Eg Annotation "co.elastic.hints.nginx/stream: stderr" will create a hints entry for container nginx
-										hints, containerProcessors = GenerateHintsForContainer(annotations, parts[0], prefix)
-									}
-								}
-							}
-						}
-					}
-				}
-			}
-		} else {
-			// If there are top level hints like co.elastic.hints/ then just add the values after the /
-			// Eg. Annotation "co.elastic.hints/stream: stderr" will will create a hints entries for all containers in the pod
-			hints = utils.GenerateHints(annotations, "", prefix)
+	cName := ""
+	cHost := ""
+
+	ann, ok := k8sMapping["annotations"]
+	if !ok {
+		return hintData
+	}
+	annotations, _ := ann.(mapstr.M)
+
+	// Get the name of the container from the metadata. We need it to extract the hints that affect it directly.
+	// E.g. co.elastic.hints.<container-name>/host: "..."
+	if con, ok := k8sMapping["container"]; ok {
+		containers, _ := con.(mapstr.M)
+		if name, err := containers.GetValue("name"); err == nil {
+			cName = name.(string)
+		}
+		if cPort, err := containers.GetValue("port"); err == nil {
+			// This is the default for the host value of a specific container.
+			cHost = "${kubernetes.pod.ip}:" + cPort.(string)
 		}
-		logger.Debugf("Extracted hints are :%v", hints)
+	}
 
-		if len(hints) > 0 {
-			hintData = GenerateHintsResult(hints, k8sMapping, annotations, logger, prefix, cID)
+	hintsExtracted := utils.GenerateHints(annotations, cName, prefix)
+	if len(hintsExtracted) == 0 {
+		return hintData
+	}
 
-			// Only if there are processors defined in a specific container we append them to the processors of the top level
-			if len(containerProcessors) > 0 {
-				hintData.processors = append(hintData.processors, containerProcessors...)
+	// Check if host exists. Otherwise, add default entry for it.
+	if cHost != "" {
+		if hintsValues, ok := hintsExtracted[hints].(mapstr.M); ok {
+			if _, ok := hintsValues[host]; !ok {
+				hintsValues[host] = cHost
+			}
+		} else {
+			hintsExtracted[hints] = mapstr.M{
+				host: cHost,
 			}
-			logger.Debugf("Generated Processors mapping :%v", hintData.processors)
 		}
 	}
 
-	return hintData
-}
-
-// Generates hints and processors list for specific containers
-func GenerateHintsForContainer(annotations mapstr.M, parts, prefix string) (mapstr.M, []mapstr.M) {
```

Comment:
We to move out of functional isolation and prefer to use a single function all-together? Having different+smaller functions  helps with readability and test-ability of the code, no?

Reply:
There was only one line on that function that was exclusive of it:
```go
containerProcessors := utils.GetConfigs(annotations, prefix, "hints."+parts+"/processors")
```

It was not worth it to have it there anymore, as most of the function that used it was refactored.